core/extension/gdnative_interface.cpp

```diff
@@ -855,6 +855,12 @@ static GDObjectInstanceID gdnative_object_get_instance_id(const GDNativeObjectPt
 	return (GDObjectInstanceID)o->get_instance_id();
 }
 
+static void gdnative_reference_assign(const GDNativeObjectPtr p_ref, const GDNativeObjectPtr p_object) {
```

Comment:
I think the signature is not correct. `p_ref` is actually a pointer to pointer, while `p_object` is a pointer.
Indeed in the assignment below we see this:
```
*(Ref<T> *)p_ref
```
Knowing Ref<T> expands to `T*`, the code would be:
```
*(T**)p_ref
```
Which means `p_ref` is not a pointer to object, it is a pointer to pointer to object.

Reply:
yeah i'm having a closer look, it doesn't compile and I was too distracted to notice :) working on it some more :)

Reply:
Note that we're not assigning the pointer, the goal is to call operator= on the reference, and I just released, that isn't exposed to take an object *, which makes sense.

Reply:
`Ref<T>::operator=` does assign a pointer though.

Reply:
It needs to do a lot more then assigning a pointer. It needs to check the current pointer that is being referenced and make sure that object is de-referenced, then assign the new pointer, then reference the object pointed to by the new pointer.

So 3 steps, and we were only doing 1 of the 3. 

The problem is that there is no operator= that takes an object pointer because within the context of Godot, this is illegal. The only scenario where you are using the pointer of a RefCounted object is on its construction. From that moment on you're only supposed to use `Ref<T>` to ensure we consistently update the refcounter.

Because we can't do this when exchanging objects between godot and godot-cpp because `Ref<T>` isn't portable between the two this fails.

Interestingly, Variant breaks this rule too so I need to have a closer look on how that works around it.

Reply:
Well I can imagine this function gets harder to write than the place we needed it since we are "further away" from the original use case in which the types of operands were pretty much known, now we have to do double-check everything which might be slower as well...

>  there is no operator= that takes an object pointer 

Can't you use `reference_ptr`?

Reply:
It's not just important to update the pointer, it's the refcount that is more important, especially from the object that may currently be there, which may be an object only known within `godot` and unknown to `godot-cpp`.